[LV][VPlan] Add initial support for CSA vectorization

michaelmaitland opened the pull request:
This patch adds initial support for CSA vectorization LLVM. This new class
can be characterized by vectorization of assignment to a scalar in a loop,
such that the assignment is conditional from the perspective of its use.
An assignment is conditional in a loop if a value may or may not be assigned
in the loop body. 

For example:

```
int t = init_val;
for (int i = 0; i < N; i++) {
  if (cond[i]) 
    t = a[i];
}
s = t; // use t
```

Using pseudo-LLVM code this can be vectorized as

```
vector.ph:
  ...
  %t = %init_val
  %init.mask = <all-false-vec>
  %init.data = <poison-vec> ; uninitialized
vector.body:
  ...
  %mask.phi = phi [%init.mask, %vector.ph], [%new.mask, %vector.body]
  %data.phi = phi [%data.mask, %vector.ph], [%new.mask, %vector.body]
  %cond.vec = <widened-cmp> ...
  %a.vec    = <widened-load> %a, %i
  %b        = <any-lane-active> %cond.vec
  %new.mask = select %b, %cond.vec, %mask.phi
  %new.data = select %b, %a.vec, %data.phi
  ...
middle.block:
  %s = <extract-last-active-lane> %new.mask, %new.data
```

On each iteration, we track whether any lane in the widened condition was active,
and if it was take the current mask and data as the new mask and data vector.
Then at the end of the loop, the scalar can be extracted only once.

This transformation works the same way for integer, pointer, and floating point
conditional assignment, since the transformation does not require inspection
of the data being assigned.

In the vectorization of a CSA, we will be introducing recipes into the vector
preheader, the vector body, and the middle block. Recipes that are introduced
into the preheader and middle block are executed only one time, and recipes
that are in the vector body will be possibly executed multiple times. The more
times that the vector body is executed, the less of an impact the preheader
and middle block cost have on the overall cost of a CSA.

A detailed explanation of the concept can be found [here](https://discourse.llvm.org/t/vectorization-of-conditional-scalar-assignment-csa/80964).

This patch is further tested in https://github.com/llvm/llvm-test-suite/pull/155.

michaelmaitland committed:
[CSA] Add CSADescriptors Analysis

michaelmaitland committed:
[LVL][CSA] Legalize CSA vectorization

michaelmaitland committed:
[LV] Build VPlan for CSA

michaelmaitland committed:
[LV] Precommit csa vectorization test cases

artagnon commented on llvm/test/Transforms/LoopVectorize/RISCV/csa.ll:
```diff
@@ -0,0 +1,5188 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py
+; RUN: opt < %s -S -mtriple riscv64 -mattr="+v" -riscv-v-vector-bits-min=256 \
+; RUN:   -passes=loop-vectorize -force-tail-folding-style=data-with-evl \
+; RUN:   -enable-csa-vectorization | FileCheck %s -check-prefix=EVL
+; RUN: opt < %s -S -mtriple riscv64 -mattr="+v" -riscv-v-vector-bits-min=256 \
+; RUN:   -passes=loop-vectorize -force-tail-folding-style=none \
+; RUN:   -enable-csa-vectorization | FileCheck %s -check-prefix=NO-EVL
+; RUN: opt < %s -S -mtriple riscv64 -mattr="+v" -riscv-v-vector-bits-min=256 \
+; RUN:   -passes=loop-vectorize -force-tail-folding-style=data \
+; RUN:   -enable-csa-vectorization | FileCheck %s -check-prefix=DATA
```
I'm not sure if it's productive to check against tail-folding-styles. Maybe force-vector-width, force-vector-interleave instead?

michaelmaitland replied:
We are generating different recipes for different tail folding styles. We should test against tail folding styles since that gives us coverage for those code paths.

artagnon commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -3803,6 +3982,20 @@ class VPlanSlp {
   /// Return true if all visited instruction can be combined.
   bool isCompletelySLP() const { return CompletelySLP; }
 };
+
+namespace vputils {
+
+/// Returns true for PHI-like recipes.
+bool isPhi(const VPRecipeBase &R);
+
+/// Returns true for PHI-like recipes that generate their own backedge
+bool isPhiThatGeneratesBackedge(const VPRecipeBase &R);
+
+/// Returns true for PHI-like recipes that exists in vector loop header basic
+/// block
+bool isHeaderPhi(const VPRecipeBase &R);
+} // end namespace vputils
+
```
Kindly move this to VPlanUtils.cpp/.h?

artagnon commented on llvm/include/llvm/Analysis/CSADescriptors.h:
```diff
@@ -0,0 +1,78 @@
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/Value.h"
+
+#ifndef LLVM_ANALYSIS_SIFIVECSADESCRIPTORS_H
+#define LLVM_ANALYSIS_SIFIVECSADESCRIPTORS_H
```
SIFIVE?

artagnon commented on llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:
```diff
@@ -1985,6 +1985,11 @@ bool RISCVTTIImpl::isLSRCostLess(const TargetTransformInfo::LSRCost &C1,
 }
 
+bool RISCVTTIImpl::enableCSAVectorization() const {
+  return ST->hasVInstructions() &&
+         ST->getProcFamily() == RISCVSubtarget::SiFive7;
```
Why are we only enabling it for SiFive7?

michaelmaitland replied:
We have measured it on a sifive7 cpu and found that it was beneficial to have enabled by default. This was not the case for all microarchitectures that we tested. I think it would be a good idea to let cpus opt into enabling this by default, at least until the feature becomes more mature.

artagnon commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -79,6 +79,10 @@ static cl::opt<LoopVectorizeHints::ScalableForceKind>
                 "Scalable vectorization is available and favored when the "
                 "cost is inconclusive.")));
 
+static cl::opt<bool>
+    EnableCSA("enable-csa-vectorization", cl::init(false), cl::Hidden,
+              cl::desc("Control whether CSA loop vectorization is enabled"));
```
Why guard it under a flag? Seems useful in the general case.

michaelmaitland replied:
This is a new feature, and I think it should be able to be toggled by a switch at least until the feature becomes more mature. This patch is disabling CSA by default (except for the sifive7 which was measured to see benefit). We can consider enabling it across the board in a future patch.

https://github.com/llvm/llvm-project/pull/83467 takes the same approach.

artagnon reviewed (commented):
I've done a very rough initial review, to get the ball rolling. Can you kindly look at iv-select-cmp.ll (and associated tests) to increase test coverage to include negative tests as well? It would be nice if this patch created changes in iv-select-cmp.ll, although I'll need to review in more detail to find out why.

artagnon commented:
Another high-level question: is it possible to not nest the test under RISCV? I realized that you've said it's disabled in BasicTTIImpl, but having a separate cas.ll under each target wouldn't be ideal, no?

michaelmaitland commented on llvm/test/Transforms/LoopVectorize/RISCV/csa.ll:
```diff
@@ -0,0 +1,5188 @@
+;       t = data[i];
+;   }
+;   return t; // use t
+; }
```
simple_csa_int_select should have been named to simple_csa_int_select_induction_cmp. But the actual body of the code in the comment is correct. The interesting thing here is the conditions:

```
// requires us to widen a loop invariant a
if (a < data[i])
// requires us to widen the induction variable i
if (i < data[i])
```

michaelmaitland commented on llvm/test/Transforms/LoopVectorize/RISCV/csa.ll:
```diff
@@ -0,0 +1,5188 @@
+  br i1 %exitcond.not, label %for.cond.cleanup, label %for.body
+}
+
+; This function is generated from the following C/C++ program:
```
This test is generated using update_test_checks script. I don't know if it is possible to get that script to generate what you recommend.

This case is harder to vectorize since it contains control flow.

artagnon replied:
I think it would be ideal if we have hand-written CHECK-NOT lines, and when updating the tests, simply `git add -p` to stage the vectorized examples and discard the hunks to the CHECK-NOT lines. That way, the tests would be less verbose, and covey the same information, no?

michaelmaitland reviewed (commented):
@artagnon thanks so much for the initial review!

I've addressed comments and made some responses :)

michaelmaitland commented:
> It would be nice if this patch created changes in iv-select-cmp.ll, although I'll need to review in more detail to find out why.

CSA vectorization is not enabled by default.

michaelmaitland commented:
>  is it possible to not nest the test under RISCV? I realized that you've said it's disabled in BasicTTIImpl, but having a separate cas.ll under each target wouldn't be ideal, no?

I can do this. It's a good idea.

artagnon commented:
> This PR contains a series of commits which together implement an initial version of conditional scalar vectorization.
> 
> A detailed explanation of the changes can be found [here](https://discourse.llvm.org/t/vectorization-of-conditional-scalar-assignment-csa/80964).

Kindly note that the only functionality is squash-and-merge, and this will be the final commit message: "series of commits" doesn't make sense in the final commit message. Also, it would be nice to have a short explanation in the commit message itself (possibly with a few examples). Finally, kindly ensure that markdown doesn't end up when someone does a `git log`: it's plain-text.

artagnon commented:
> > is it possible to not nest the test under RISCV? I realized that you've said it's disabled in BasicTTIImpl, but having a separate cas.ll under each target wouldn't be ideal, no?
> 
> I can do this. It's a good idea.

Perhaps a good idea to name the test conditional-scalar-assignment.ll instead of the more cryptic csa.ll.

michaelmaitland commented:
I have updated the tests to use `-NOT: vector.body` for the cases that cannot be vectorized
I have also moved tests up a directory. I had to keep a copy of some in the RISCV target directory since they depend on TTI specific to RISC-V (i.e. pointer size).

artagnon commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -3803,6 +3982,7 @@ class VPlanSlp {
   /// Return true if all visited instruction can be combined.
   bool isCompletelySLP() const { return CompletelySLP; }
 };
+
```
Stray change?

artagnon commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -668,6 +672,94 @@ Value *VPInstruction::generatePerPart(VPTransformState &State, unsigned Part) {
+      State.set(this, InitData, Part);
+      return InitData;
+    }
+    Value *V = State.get(this, Part - 1);
+    return V;
```
`return State.get(this, Part - 1)`, and above/below?

artagnon commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -668,6 +672,94 @@ Value *VPInstruction::generatePerPart(VPTransformState &State, unsigned Part) {
+    BasicBlock *PreheaderBB = State.CFG.getPreheaderBBFor(this);
+
+    // InitVL can be anything since it won't be used if no mask was active
+    Value *InitVL = ConstantInt::get(State.Builder.getInt32Ty(), 0);
```
How was int32 chosen?

michaelmaitland replied:
According to the [LangRef](https://llvm.org/docs/LangRef.html#vector-predication-intrinsics):

> The explicit vector length parameter always has the type i32 and is an unsigned integer value. The explicit vector length parameter (%evl) is in the range:

artagnon commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -676,11 +768,16 @@ Value *VPInstruction::generatePerPart(VPTransformState &State, unsigned Part) {
+  return getOpcode() == VPInstruction::ResumePhi ||
+         getOpcode() == VPInstruction::CSAVLPhi ||
+         getOpcode() == VPInstruction::CSAVLSel ||
+         getOpcode() == VPInstruction::ExplicitVectorLength;
```
Is `VPInstruction::ExplicitVectorLength` something that this patch adds? If not, how come it appears in the diff?

michaelmaitland replied:
I created a separate patch for this:

https://github.com/llvm/llvm-project/pull/106818

artagnon commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -1998,6 +2119,223 @@ void VPScalarCastRecipe ::print(raw_ostream &O, const Twine &Indent,
+    // inactive, which would also cause the reduction to have value 0.
+    Value *MaybeLastIdx = State.Builder.CreateIntMaxReduce(ActiveLaneIdxs);
+    Value *IsLaneZeroActive =
+        State.Builder.CreateExtractElement(MaskSel, (uint64_t)0);
```
Prefer `static_cast` over C-style cast.

artagnon commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1435,6 +1439,51 @@ bool VPlanTransforms::tryAddExplicitVectorLength(VPlan &Plan) {
+    VPInstruction *VPEVL, const MapVector<PHINode *, VPCSAState *> &CSAStates,
+    bool PlanHasScalarVFOnly) {
+
+  // We build the scalar version of a CSA when VF=ElementCount::getFixed(1),
+  // which does not require an EVL.
+  if (PlanHasScalarVFOnly)
+    return;
```
Is this a planned extension? I'm not sure what you can do with VF=1, so maybe drop `PlanHasScalarVFOnly` argument and handle in caller?

artagnon commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1435,6 +1439,51 @@ bool VPlanTransforms::tryAddExplicitVectorLength(VPlan &Plan) {
+  if (PlanHasScalarVFOnly)
+    return;
+
+  for (auto CSA : CSAStates) {
```
`auto &[_, CSAState] : CSAStates`?

artagnon commented on llvm/lib/Transforms/Vectorize/VPlanUtils.cpp:
```diff
@@ -60,3 +60,30 @@ bool vputils::isHeaderMask(const VPValue *V, VPlan &Plan) {
          IsWideCanonicalIV(A) && B == Plan.getOrCreateBackedgeTakenCount();
 }
+
+bool vputils::isPhi(const VPRecipeBase &R) {
+  if (R.isPhi())
+    return true;
+  if (auto *VPInst = dyn_cast<VPInstruction>(&R))
+    return VPInst->getOpcode() == VPInstruction::CSAMaskPhi ||
+           VPInst->getOpcode() == VPInstruction::CSAVLPhi;
+  return false;
+}
+
+bool vputils::isPhiThatGeneratesBackedge(const VPRecipeBase &R) {
+  if (isa<VPWidenPHIRecipe, VPCSAHeaderPHIRecipe>(&R))
+    return true;
+  if (auto *VPInst = dyn_cast<VPInstruction>(&R))
+    return VPInst->getOpcode() == VPInstruction::CSAMaskPhi ||
+           VPInst->getOpcode() == VPInstruction::CSAVLPhi;
+  return false;
+}
+
+bool vputils::isHeaderPhi(const VPRecipeBase &R) {
+  if (isa<VPHeaderPHIRecipe, VPWidenPHIRecipe>(&R))
+    return true;
+  if (auto *VPInst = dyn_cast<VPInstruction>(&R))
+    return VPInst->getOpcode() == VPInstruction::CSAMaskPhi ||
+           VPInst->getOpcode() == VPInstruction::CSAVLPhi;
+  return false;
```
Hm, not a fan of this code. I wonder if we can do something to automatically let `CSAVLPhi::isPhi` return true?

michaelmaitland replied:
I have refactored this code

artagnon commented on llvm/test/Transforms/LoopVectorize/conditional-scalar-assignment.ll:
```diff
@@ -0,0 +1,3091 @@
+  %spec.select.lcssa = phi i32 [ %spec.select, %for.body ]
+  br label %for.cond.cleanup
+
+for.cond.cleanup:                                 ; preds = %for.cond.cleanup.loopexit, %entry
+  %t.0.lcssa = phi i32 [ -1, %entry ], [ %spec.select.lcssa, %for.cond.cleanup.loopexit ]
+  ret i32 %t.0.lcssa
```
Tests could be made a little better, but I know it's a large test, and we don't have any tools to refactor .ll files. Example: s/for.cond.cleanup/exit/, lcssa is unnecessary, s/indvars.iv/iv/.

michaelmaitland replied:
Are you sure that LCSSA is not necessary? `void InnerLoopVectorizer::fixCSALiveOuts(VPTransformState &State, VPlan &Plan)` exists to fix lcssa phis. LCSSA form that is generated into the vectorizer from clang.

artagnon replied:
IIUC, I don't think it's necessary. See this test in iv-select-cmp.ll for instance:

```llvm
define i64 @select_icmp_const_1(ptr %a, i64 %n) {
; CHECK-LABEL: define i64 @select_icmp_const_1
; CHECK-NOT:   vector.body:
;
entry:
  br label %for.body

for.body:                                         ; preds = %entry, %for.body
  %iv = phi i64 [ %inc, %for.body ], [ 0, %entry ]
  %rdx = phi i64 [ %cond, %for.body ], [ 3, %entry ]
  %arrayidx = getelementptr inbounds i64, ptr %a, i64 %iv
  %0 = load i64, ptr %arrayidx, align 8
  %cmp2 = icmp eq i64 %0, 3
  %cond = select i1 %cmp2, i64 %iv, i64 %rdx
  %inc = add nuw nsw i64 %iv, 1
  %exitcond.not = icmp eq i64 %inc, %n
  br i1 %exitcond.not, label %exit, label %for.body

exit:                                             ; preds = %for.body
  ret i64 %cond
}
```

michaelmaitland replied:
I agree that LCSSA isn't required to vectorize a CSA. However, the code that clang generates is in LCSSA (the IR in the tests comes from C code with the -stop-before=loop-vectorize), so without it we do not have coverage on code that **actually** gets generated. I'm not sure its such a good idea to remove test coverage for IR that we know will be generated -- simplifying it does not fully test CSA since we'd be skipping the `fixCSAPhis` function.

artagnon replied:
Makes sense. Do keep in mind, however, that clang isn't the only frontend to LLVM, and other frontends (rust, swift etc) might not be running the optimization pipeline the same way clang is: do ensure coverage of those codepaths as well.

michaelmaitland replied:
> Do keep in mind, however, that clang isn't the only frontend to LLVM, and other frontends (rust, swift etc) might not be running the optimization pipeline the same way clang is: do ensure coverage of those codepaths as well.

That is true. The way the tests are "most inclusive" since we're going out of the way to test the extra case of LCSSA. Maybe it would make sense to have only a few in LCSSA to make sure we test that path, and simplify the rest. WDYT?

artagnon replied:
I think this would be wise, as reduced cases tend to catch more bugs. It would be in line with the rest of the testing in LoopVectorize/test.

michaelmaitland replied:
I have pushed a commit that reduces test cases.

artagnon reviewed (commented):
I did a closer inspection of the patch, and found mostly small things to fix. Otherwise, I'm a fan of the approach you've taken: the code is quite easy to understand, and there is little bloat. Thanks again for the patch, and I hope the other reviewers also have similar sentiments!

On a minor note, it might be helpful if you could locally enable CSA vectorization, and post the diff of the test changes somewhere, just to make sure we're doing the right thing.

michaelmaitland commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -866,14 +879,23 @@ bool LoopVectorizationLegality::canVectorizeInstrs() {
 
+        // Check if the PHI can be classified as a CSA PHI.
+        if (EnableCSA || (TTI->enableCSAVectorization() &&
+                          EnableCSA.getNumOccurrences() == 0)) {
```
https://www.llvm.org/doxygen/classllvm_1_1cl_1_1Option.html#a2dc75fafb68e7669e776e19da7857cdf

This if statement has the following semantics: 
* True if EnableCSA was true by default (it isn't currently)
* True if `enable-csa-vectorization=true` was passed
* True if the subtarget enables CSA by default and it `-enable-csa-vectorization=false` wasn't passed

This gives us the following behaviors:
* Take whatever enable-csa-vectorization is set as if it was passed
* Else defer to subtarget to tell us whether they want csa enabled by default

michaelmaitland commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -2932,6 +2938,25 @@ LoopVectorizationCostModel::getVectorIntrinsicCost(CallInst *CI,
+    Value *ExtractedScalar = State.get(CSA.second->getExtractScalarRecipe(), 0,
+                                       /*NeedsScalar=*/true);
+    // Fix LCSSAPhis
+    llvm::SmallPtrSet<PHINode *, 2> ToFix;
+    for (User *U : V->users())
+      if (auto *Phi = dyn_cast<PHINode>(U);
+          Phi && Phi->getParent() == LoopExitBlock)
+        ToFix.insert(Phi);
+    for (PHINode *Phi : ToFix)
```
It causes a crash if we do it in place, likely due to modifying `V->users()` while iterating over it. I'll take a closer look on Monday.

michaelmaitland commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -8540,6 +8599,107 @@ VPRecipeBuilder::tryToCreateWidenRecipe(Instruction *Instr,
+        CSA.first) {
+      auto *VPNotCond = new VPInstruction(VPInstruction::Not, WidenedCond, DL);
+      VPNotCond->insertBefore(
+          GetVPValue(CSA.second.getAssignment())->getDefiningRecipe());
```
Updated to use VPBuilder to create vp instructions. used builder to insert where appropriate.

ElvisWang123 commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -2107,6 +2241,223 @@ void VPScalarCastRecipe ::print(raw_ostream &O, const Twine &Indent,
+  // FIXME: These costs should be moved into VPInstruction::computeCost. We put
+  // them here for now since they are related to updating the data and there is
+  // no VPInstruction::computeCost support at the moment. CSAInitMask AnyActive
+  C += TTI.getArithmeticInstrCost(Instruction::Select, VTy, CostKind);
```
We should use `TTI.getCmpSelInstrCost()` to get the instruction cost of `Select` instructions.

ElvisWang123 commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -2107,6 +2241,223 @@ void VPScalarCastRecipe ::print(raw_ostream &O, const Twine &Indent,
+  // VPVLSel
+  C += TTI.getArithmeticInstrCost(Instruction::Select, VTy, CostKind);
+  // MaskUpdate
+  C += TTI.getArithmeticInstrCost(Instruction::Select, MaskTy, CostKind);
```
And here.

ElvisWang123 commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -2107,6 +2241,223 @@ void VPScalarCastRecipe ::print(raw_ostream &O, const Twine &Indent,
+                                    CostKind);
+    // IsLaneZeroActive
+    C += TTI.getArithmeticInstrCost(Instruction::ExtractElement, MaskTy,
+                                    CostKind);
```
We should use `TTI.getVectorInstrCost()` to get the instruction cost of `ExtractElement` instructions.

ElvisWang123 commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -2107,6 +2241,223 @@ void VPScalarCastRecipe ::print(raw_ostream &O, const Twine &Indent,
+                                    CostKind);
+    // MaybeLastIdxEQZero
+    C += TTI.getArithmeticInstrCost(Instruction::ICmp, MaskTy->getScalarType(),
+                                    CostKind);
```
We should use `TTI.getCmpSelInstrCost()` to get the instruction cost of `Cmp` instructions.

ElvisWang123 commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -2107,6 +2241,223 @@ void VPScalarCastRecipe ::print(raw_ostream &O, const Twine &Indent,
+                                    CostKind);
+    // LastIdx
+    C += TTI.getArithmeticInstrCost(Instruction::Select, VTy->getScalarType(),
+                                    CostKind);
```
And here.

ElvisWang123 commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -2107,6 +2241,223 @@ void VPScalarCastRecipe ::print(raw_ostream &O, const Twine &Indent,
+  // LastIdxGeZero
+  C += TTI.getArithmeticInstrCost(Instruction::ICmp, Int32VTy, CostKind);
+  // ChooseFromVecOrInit
+  C += TTI.getArithmeticInstrCost(Instruction::Select, VTy->getScalarType(),
```
And here.

michaelmaitland commented:
@ElvisWang123 thank you for your review. I think I have addressed all of your comments.

ElvisWang123 commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -2107,6 +2241,225 @@ void VPScalarCastRecipe ::print(raw_ostream &O, const Twine &Indent,
+                                CmpInst::BAD_ICMP_PREDICATE, CostKind);
+  }
+  // ExtractFromVec
+  C += TTI.getArithmeticInstrCost(Instruction::ExtractElement, VTy, CostKind);
```
Sorry I missed this before. We should use `getVectorInstrCost()` here.

michaelmaitland replied:
Thanks! updated.

michaelmaitland commented:
Rebased main.

ElvisWang123 commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -2139,6 +2273,225 @@ void VPScalarCastRecipe ::print(raw_ostream &O, const Twine &Indent,
+  const TargetTransformInfo &TTI = Ctx.TTI;
+
+  // Data Update
+  C += TTI.getArithmeticInstrCost(Instruction::Select, VTy, CostKind);
```
We should use `getCmpSelInstrCost()` here. Sorry I missed this.

michaelmaitland replied:
i am sorry i missed it too 😆

artagnon commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1313,6 +1313,45 @@ void VPlanTransforms::addActiveLaneMask(
+    // extract the scalar from the data vector, we must use the EVL that
+    // corresponds to the EVL that was used when the mask vector was last
+    // updated. To do this, we introduce CSAVLPhi and CSAVLSel instructions
+    auto *VPVLPhi =
+        new VPInstruction(VPInstruction::CSAVLPhi, {}, {}, "csa.vl.phi");
+    auto *VPVLSel =
+        new VPInstruction(VPInstruction::CSAVLSel,
+                          {VPAnyActiveEVL, VPVLPhi, &EVL}, {}, "csa.vl.sel");
+    VPVLPhi->insertAfter(CSAState->getPhiRecipe());
+    VPVLSel->insertAfter(VPAnyActiveEVL);
```
Can use VPBuilder for these as well?

artagnon commented on llvm/test/Transforms/LoopVectorize/conditional-scalar-assignment.ll:
```diff
@@ -0,0 +1,2415 @@
+;   return t; // use t
+; }
+define i32 @simple_csa_int(i32 %N, ptr %cond, ptr %data) {
+; EVL-LABEL: @simple_csa_int(
+; EVL-NOT: vector.body:
+;
+; NO-EVL-LABEL: @simple_csa_int(
+; NO-EVL-NOT: vector.body:
+;
```
You can use `--check-prefixes=CHECK,EVL` and `--check-prefixes=CHECK,NO-EVL` to have these as a common `CHECK-NOT:`.

artagnon commented on llvm/test/Transforms/LoopVectorize/conditional-scalar-assignment.ll:
```diff
@@ -0,0 +1,2415 @@
+  %cmp19 = icmp sgt i32 %N, 0
+  br i1 %cmp19, label %for.body.preheader, label %exit
+
+for.body.preheader:                               ; preds = %entry
+  %wide.trip.count = zext i32 %N to i64
+  br label %for.body
+
+exit.loopexit:                        ; preds = %for.body
+  %0 = fadd float %t.1, %s.1
+  br label %exit
+
+exit:                                 ; preds = %exit.loopexit, %entry
+  %add = phi float [ %0, %exit.loopexit ], [ 2.000000e+00, %entry ]
+  ret float %add
+
+for.body:                                         ; preds = %for.body.preheader, %for.body
```
nit: s/for.body/loop/, s/for.body.preheader/loop.ph/. Can `exit.loopexit` be merged into `exit`?

michaelmaitland replied:
> Can exit.loopexit be merged into exit

I don't think it makes sense. Moving the `fadd` into `exit` is weird. What should the phi values be coming from entry? And we don't want to add an unconditional add into `exit` if we don't enter the loop.

artagnon commented on llvm/lib/Transforms/Vectorize/LoopVectorizationPlanner.h:
```diff
@@ -164,8 +164,8 @@ class VPBuilder {
   }
-  VPValue *createNot(VPValue *Operand, DebugLoc DL = {},
-                     const Twine &Name = "") {
+  VPInstruction *createNot(VPValue *Operand, DebugLoc DL = {},
+                           const Twine &Name = "") {
     return createInstruction(VPInstruction::Not, {Operand}, DL, Name);
   }
```
@fhahn: see also this, for #108858.

artagnon commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -1564,11 +1586,15 @@ bool LoopVectorizationLegality::canFoldTailByMasking() const {
   // TODO: handle non-reduction outside users when tail is folded by masking.
   for (auto *AE : AllowedExit) {
     // Check that all users of allowed exit values are inside the loop or
-    // are the live-out of a reduction.
```
Very minor nit: period ending sentence dropped?

artagnon reviewed (commented):
Thanks, the code looks much better now!

michaelmaitland commented:
I have rebased and responded to the last round of reviews.

In the last round of reviews, I found that the data-with-evl cases were not getting handled as expected. Part of this was because I was missing `-prefer-predicate-over-epilogue=predicate-dont-vectorize` which overrides `data-with-evl`. I have added that option.

Once that option was added, I realized that we were unable to vectorize. Thats because the generic target does not support scalable vectorization nor predicated vectroization. We can enable scalable vectorization with `-scalable-vectorization=on`, but there is no way to make a generic target support predicated vectorization to my knowledge. As a result, I have moved these tests back into RISC-V. I think we can consider moving them back out until a generic target can handle it.

michaelmaitland commented:
ping

michaelmaitland reviewed (commented):
@huntergr-arm thanks for the review. I have addressed or responded to your comments.

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -2914,6 +2920,25 @@ LoopVectorizationCostModel::getVectorIntrinsicCost(CallInst *CI,
+    assert(VPDataUpdate &&
+           "VPDataUpdate must have been introduced prior to fixing live outs");
+    Value *V = VPDataUpdate->getUnderlyingValue();
+    Value *ExtractedScalar =
```
Can those fixups be directly done in VPlan, similar to how other live-outs are handled (with the exception of induction users, which is the last remaining case that needs fixing up outside of VPlan)?

michaelmaitland replied:
addressed

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -669,6 +673,65 @@ Value *VPInstruction::generate(VPTransformState &State) {
   }
+  case VPInstruction::CSAMaskPhi: {
+    IRBuilder<>::InsertPointGuard Guard(State.Builder);
+    State.Builder.SetInsertPoint(State.CFG.PrevBB->getFirstNonPHI());
```
Why does the builder need to be reset here?

michaelmaitland replied:
addressed

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -3497,6 +3695,8 @@ class VPlan {
   /// live-outs are fixed via VPLiveOut::fixPhi.
   MapVector<PHINode *, VPLiveOut *> LiveOuts;
 
+  MapVector<PHINode *, VPCSAState *> CSAStates;
```
Why does this need to be part of VPlan? Is it not possible to encode all required info in recipes?

michaelmaitland replied:
removed

fhahn commented on llvm/test/Transforms/LoopVectorize/RISCV/conditional-scalar-assignment.ll:
```diff
@@ -0,0 +1,2932 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 5
```
Can the general code-gen parts be tested independent of RISCV? Could you add a test checking the printed VPlan containing the CAS recipes?

michaelmaitland replied:
There is no way to make a generic target support predicated vectorization to my knowledge. As a result, I have moved these tests back RISC-V. I think we can consider moving them back out until a generic target can handle it. We do something similar for other code that uses this option. WDYT?

fhahn replied:
I think I am missing something, but the non-EVL codegen doens't seem to use and predicated memory accesses, so it should be able to do it with a generic target?

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -231,6 +231,53 @@ class VPLane {
+/// VPCSAState holds information required to vectorize a conditional scalar
+/// assignment.
+class VPCSAState {
+  VPValue *VPInitScalar = nullptr;
```
Would be  good to avoid holding references to VPValues/Recipes, as they may be removed/replaced, in case the state here gets stale.

michaelmaitland replied:
removed

Mel-Chen commented:
> @Mel-Chen 's #67812 seems to be quite similar work which just extends the existing code. This CSA patch is more general in that it can handle values that aren't just the induction variable, but I think the amount of extra code can be cut down a little.

#67812 only handles increasing induction variables because `FindLastIV` was designed to address the index reduction in min/max with index reductions, and the index is typically an increasing induction variable.

Currently, I do not have plans to support `FindLast`, mainly because RISC-V has `vfirst` but no `vlast` instruction. However, I can still provide some suggestions.
For idiom recognition, we could directly extend the `RecurrenceDescriptor`. 
For vectorization, I recommend focusing on an in-loop reduction approach for this idiom, as an out-loop reduction would require generating additional recurrence phi in the vectorized loop, making vectorization too costly. To implement in-loop reduction, it's best to introduce a new intrinsic, `vector.extractlast(%scalar_start, %vec_val, %vec_mask)`, with semantics as follows: 
```
  %c = vcpop(%vec_mask)
  if(%c == 0)
    return %scalar_start;
  %vec_com = compress(%vec_val, %vec_mask)
  return extractelement %vec_com, %c - 1
```
For targets without a corresponding instruction, such as RISC-V, we could implement the same semantics in the backend using multiple instructions (e.g., vcpop + beqz + vcompress) if it is profitable. In VPlan, we might need to create a new recipe, `VPExtractLastRecipe`, to emit this intrinsic. I hope these suggestions are helpful.

In the end, if there is indeed a need to support this semantic, please let me know, and I can discuss internally whether to include it in the plan. :)

michaelmaitland commented:
rebase; ping

michaelmaitland commented:
ping

michaelmaitland commented:
ping!

michaelmaitland commented:
rebased

michaelmaitland commented:
@fhahn @ayalz ping

ElvisWang123 commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -2346,6 +2444,222 @@ void VPScalarCastRecipe ::print(raw_ostream &O, const Twine &Indent,
+    return 0;
+
+  InstructionCost C = 0;
+  auto *VTy = VectorType::get(getUnderlyingValue()->getType(), VF);
```
Using `inferScalarType(this)` instead of `getUnderlyingValue()->getType()` would be better.
And I think you also need to implement `inferScalarType()` in `VPlanAnalysis` for these CSA recipes.

ElvisWang123 commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -2346,6 +2444,222 @@ void VPScalarCastRecipe ::print(raw_ostream &O, const Twine &Indent,
+    return 0;
+
+  InstructionCost C = 0;
+  auto *VTy = VectorType::get(getUnderlyingValue()->getType(), VF);
```
Using `inferScalarType(this)` instead of `getUnderlyingValue()->getType()` would be better.

ElvisWang123 commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -2346,6 +2444,222 @@ void VPScalarCastRecipe ::print(raw_ostream &O, const Twine &Indent,
+    return 0;
+
+  InstructionCost C = 0;
+  auto *VTy = VectorType::get(getUnderlyingValue()->getType(), VF);
```
Using `inferScalarType(this)` instead of `getUnderlyingValue()->getType()` would be better.

michaelmaitland commented:
ping! @fhahn @ayalz

fhahn commented on llvm/include/llvm/Transforms/Vectorize/LoopVectorizationLegality.h:
```diff
@@ -550,6 +560,10 @@ class LoopVectorizationLegality {
   void addInductionPhi(PHINode *Phi, const InductionDescriptor &ID,
                        SmallPtrSetImpl<Value *> &AllowedExit);
 
+  // Updates the vetorization state by adding \p Phi to the CSA list.
```
```suggestion
/// Updates the vetorization state by adding \p Phi to the CSA list.
```

fhahn commented on llvm/include/llvm/Analysis/IVDescriptors.h:
```diff
@@ -423,6 +424,61 @@ class InductionDescriptor {
 
+/// A Conditional Scalar Assignment (CSA) is an assignment from an initial
+/// scalar that may or may not occur.
+class CSADescriptor {
```
I don't think CAS is a very common term, would be good to have a more descriptive name if possible

fhahn commented on llvm/lib/Analysis/IVDescriptors.cpp:
```diff
@@ -1570,3 +1571,58 @@ bool InductionDescriptor::isInductionPHI(
+/// that describes an InvalidCSA.
+bool CSADescriptor::isCSAPhi(PHINode *Phi, Loop *TheLoop, CSADescriptor &CSA) {
+
+  // Must be a scalar
```
```suggestion
  // Must be a scalar.
```

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -7550,9 +7555,17 @@ InstructionCost LoopVectorizationPlanner::cost(VPlan &Plan,
+  // preheader and middle block. It also contains recipes that are not backed by
+  // underlying instructions in the original loop. This makes it difficult to
+  // model in the legacy cost model.
+  if (!Legal.getCSAs().empty())
```
Better check for the CAS recipes?

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -8998,6 +9108,16 @@ static SetVector<VPIRInstruction *> collectUsersInExitBlocks(
           if (ExitVPBB->getSinglePredecessor() == MiddleVPBB)
             continue;
         }
+        // Exit values for CSAs are computed and updated outside of VPlan and
```
I am missing where the exit values are computed outside of VPlan. What's preventing them from being competed in VPlan?

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -649,6 +652,57 @@ Value *VPInstruction::generate(VPTransformState &State) {
+
+    Value *StartValue =
+        ConstantInt::get(WidenedCond->getType()->getScalarType(), 0);
+    Value *AnyOf = State.Builder.CreateIntrinsic(
```
Better use VPWidenIntrinsic like for other EVL based recipes if possible?

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -649,6 +652,57 @@ Value *VPInstruction::generate(VPTransformState &State) {
+    Value *VLPhi = State.get(getOperand(1), /*NeedsScalar=*/true);
+    Value *EVL = State.get(getOperand(2), /*NeedsScalar=*/true);
+    Value *VLSel = State.Builder.CreateSelect(AnyOf, EVL, VLPhi, "csa.vl.sel");
+    cast<PHINode>(VLPhi)->addIncoming(VLSel, State.CFG.PrevBB);
```
Why do we have to update the IR for one of the operands here? Can we not update the incoming values like for other phi recipes?

Then we could use regular select?

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1569,6 +1622,11 @@ static void transformRecipestoEVLRecipes(VPlan &Plan, VPValue &EVL) {
+
+  // We build the scalar version of a CSA when VF=ElementCount::getFixed(1),
+  // which does not require an EVL.
+  if (!Plan.hasScalarVFOnly())
```
does this apply to the other EVL recipes as well?

fhahn reviewed (commented):
I added some comments inline.

On the high level, it would be good to have a test that checks the printed VPlan, to see things on a higher level.

Supporting both EVL and non-EVL in a single patch makes the patch quite big and I think it would help to separate them and first start with the simpler case

michaelmaitland commented:
Closing in favor of smaller patches. First one is posted at #121222
